src/datatable/datatable.jl

```diff
@@ -74,32 +74,41 @@ type DataTable <: AbstractDataTable
     colindex::Index
 
     function DataTable(columns::Vector{Any}, colindex::Index)
-        ncols = length(columns)
-        if ncols > 1
-            nrows = length(columns[1])
-            equallengths = true
-            for i in 2:ncols
-                equallengths &= length(columns[i]) == nrows
-            end
-            if !equallengths
-                msg = "All columns in a DataTable must be the same length"
-                throw(ArgumentError(msg))
-            end
+        if length(columns) == length(colindex) == 0
+            return new(Vector{Any}(0), Index())
+        elseif length(columns) != length(colindex)
+            throw(DimensionMismatch("Number of columns and column names are different"))
         end
-        if length(colindex) != ncols
-            msg = "Columns and column index must be the same length"
-            throw(ArgumentError(msg))
+        lengths = length.(columns)
+        minlen, maxlen = extrema(lengths)
+        if minlen == 0 && maxlen == 0
+            return new(columns, colindex)
+        elseif (minlen == 0 && maxlen > 0) || any(x -> x != 0, mod(maxlen, lengths))
+            throw(DimensionMismatch("Incompatible lengths of arguments"))
+        else
+            for i in 1:length(columns)
+                if isa(columns[i], Range)
+                    columns[i] = collect(columns[i])
+                end
+                repeats = div(maxlen, length(columns[i]))
```

Comment:
I thought of this mainly because I saw it in the tests
https://github.com/JuliaData/DataTables.jl/blob/5f611bf480632507e15bb2c96bf39aedb71d6d06/src/groupeddatatable/grouping.jl#L65
https://github.com/JuliaData/DataTables.jl/blob/20c71d6d40b3b238e902189b8262ba2b2e679b31/src/abstractdatatable/reshape.jl#L64
https://github.com/JuliaData/DataTables.jl/blob/20c71d6d40b3b238e902189b8262ba2b2e679b31/src/subdatatable/subdatatable.jl#L74
https://github.com/JuliaData/DataTables.jl/blob/5f611bf480632507e15bb2c96bf39aedb71d6d06/test/grouping.jl#L6
https://github.com/JuliaData/DataTables.jl/blob/20c71d6d40b3b238e902189b8262ba2b2e679b31/test/data.jl#L146
https://github.com/JuliaData/DataTables.jl/blob/20c71d6d40b3b238e902189b8262ba2b2e679b31/src/abstractdatatable/join.jl#L293

Of course, the shortcoming here is that we decide for the user which is the default, inner or outer repeat. But at least one of these DataTable constructor behaviors will have a smaller declaration